### PR TITLE
Pass context to node interface

### PR DESCRIPTION
**Pull request by dennwc**
Requires https://github.com/bblfsh/libuast/pull/83.

Node interface defined in `libuast` does not allow to properly control Node lifetime in client languages with GC.

For example, if client decides to pass an opaque handles to `libuast` instead of passing raw native pointers, it will need to know when exactly the lifetime of allocated objects ends. This is possible if the client can associate the `Uast*` with nodes that it accessed, and deallocate all the objects associated with this `Uast*` before freeing it.

The PR adds the `const Uast*`  as a first argument to all node interface functions. This allows client to associate this pointer with allocated nodes and destroy them before calling `UastFree`.

### CODE REVIEW

File: src/uast.h [hunk 1]
```diff
@@ -1,20 +1,70 @@
...
-#include "node_iface.h"
-#include "nodes.h"
-
 // Uast stores the general context required for library functions.
 // It must be initialized with `UastNew` passing a valid implementation of the
 // `NodeIface` interface.
 // Once it is not used anymore, it shall be released calling `UastFree`.
 typedef struct Uast Uast;
 
+typedef uintptr_t NodeHandle;
```
Comment (juanjux) [inline src/uast.h]: Why was this moved from `node_iface.h`?

Comment (dennwc) [reply src/uast.h]: Because `node_iface.h` requires `Uast*` that is defined in `uast.h`, and `uast.h` requires `NodeIface`.
